raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1017,7 +1017,16 @@ private FetchResponseData tryCompleteFetchRequest(
             long fetchOffset = request.fetchOffset();
             int lastFetchedEpoch = request.lastFetchedEpoch();
             LeaderState<T> state = quorum.leaderStateOrThrow();
-            ValidOffsetAndEpoch validOffsetAndEpoch = log.validateOffsetAndEpoch(fetchOffset, lastFetchedEpoch);
+
+            Optional<OffsetAndEpoch> latestSnapshotId = log.latestSnapshotId();
+            final ValidOffsetAndEpoch validOffsetAndEpoch;
+            if (fetchOffset == 0 && latestSnapshotId.isPresent()) {
```

Comment:
I have tried to figure out some other cases we prefer to read the snapshot, a common case is `highWatermark` - `fetchOffset` > count(snapshot record). Do you think it's possible to write count(snapshot record) in the snapshot header?

Reply:
Thanks @dengziming 

We had a similar conversation in another PR: https://github.com/apache/kafka/pull/13834#discussion_r1224779841

In short it is not clear to me that these improvements (implementation complexities) are a big win for the cluster metadata partition.